diff --git a/CMakeLists.txt b/CMakeLists.txt
index 05a94d3..8a6888f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -86,6 +86,10 @@ if(BUILD_TESTING)
    set(BUILD_TOOLS TRUE)
 endif()
 
+if (CXX_STDLIB)
+  set (CXX_STDLIB_FLAGS "-stdlib=${CXX_STDLIB}")
+endif ()
+
 set(SMI_MIN_VERSION "0.20")
 find_package(SharedMimeInfo ${SMI_MIN_VERSION} REQUIRED)
 
diff --git a/cmake/modules/FindStdlibInclude.cmake b/cmake/modules/FindStdlibInclude.cmake
index 135e655..0a1915a 100644
--- a/cmake/modules/FindStdlibInclude.cmake
+++ b/cmake/modules/FindStdlibInclude.cmake
@@ -8,7 +8,7 @@ function(findStdlibInclude includeName filePath)
     )
     # With the "-v" flag, compiler will print out list of default include paths
     # (i.e. stdlib include paths)
-    execute_process(COMMAND ${CMAKE_CXX_COMPILER} -Wp,-v -E ${CMAKE_CURRENT_BINARY_DIR}/empty_test.cpp
+    execute_process(COMMAND ${CMAKE_CXX_COMPILER} ${CXX_STDLIB_FLAGS} -Wp,-v -E ${CMAKE_CURRENT_BINARY_DIR}/empty_test.cpp
                     OUTPUT_QUIET
                     ERROR_VARIABLE compilerSearchPaths
     )
