--- a/CMakeLists.txt	2011-03-17 01:21:07.000000000 +0100
+++ b/CMakeLists.txt	2016-01-22 01:43:37.148109260 +0100
@@ -163,7 +163,7 @@
   SET(BUILD_CONTRIBS_LIB 1)
 ENDIF ( BUILD_CONTRIBS )
 IF ( BUILD_CONTRIBS_LIB )
-  ADD_SUBDIRECTORY (src/contribs-lib EXCLUDE_FROM_ALL)
+  ADD_SUBDIRECTORY (src/contribs-lib)
 ENDIF ( BUILD_CONTRIBS_LIB )
 
 
--- a/src/contribs-lib/CMakeLists.txt	2011-03-17 01:21:07.000000000 +0100
+++ b/src/contribs-lib/CMakeLists.txt	2016-01-22 01:43:37.148109260 +0100
@@ -106,9 +106,26 @@
 )
 TARGET_LINK_LIBRARIES(clucene-contribs-lib ${clucene_contrib_extra_libs})
 
+#install public headers.
+FOREACH(file ${HEADERS})
+	get_filename_component(apath ${file} PATH)
+	get_filename_component(aname ${file} NAME)
+	file(RELATIVE_PATH relpath ${CMAKE_SOURCE_DIR}/src/contribs-lib ${apath})
+	IF ( NOT aname MATCHES "^_.*" )
+		install(FILES ${file} 
+		        DESTINATION include/${relpath}
+		        COMPONENT development)
+	ENDIF ( NOT aname MATCHES "^_.*" )
+ENDFOREACH(file)
+
 #set properties on the libraries
 SET_TARGET_PROPERTIES(clucene-contribs-lib PROPERTIES
     VERSION ${CLUCENE_VERSION}
     SOVERSION ${CLUCENE_SOVERSION}
     COMPILE_DEFINITIONS_DEBUG _DEBUG
 )
+
+#and install library
+install(TARGETS clucene-contribs-lib 
+        DESTINATION ${LIB_DESTINATION}  
+        COMPONENT runtime )
--- a/src/shared/CLucene/LuceneThreads.h	2011-03-17 01:21:07.000000000 +0100
+++ b/src/shared/CLucene/LuceneThreads.h	2016-01-22 01:43:30.450975553 +0100
@@ -7,6 +7,7 @@
 #ifndef _LuceneThreads_h
 #define  _LuceneThreads_h
 
+#include <pthread.h>
 
 CL_NS_DEF(util)
 class CLuceneThreadIdCompare;
