--- a/channels/client/CMakeLists.txt	2014-09-12 00:46:32.000000000 +0200
+++ b/channels/client/CMakeLists.txt	2015-09-22 10:49:35.143095353 +0200
@@ -24,7 +24,9 @@
 	${CMAKE_CURRENT_SOURCE_DIR}/addin.c
 	${CMAKE_CURRENT_SOURCE_DIR}/addin.h)
 
+if(CHANNEL_STATIC_CLIENT_ENTRIES)
 list(REMOVE_DUPLICATES CHANNEL_STATIC_CLIENT_ENTRIES)
+endif()
 
 foreach(STATIC_ENTRY ${CHANNEL_STATIC_CLIENT_ENTRIES})
 	foreach(STATIC_MODULE ${CHANNEL_STATIC_CLIENT_MODULES})
--- a/channels/CMakeLists.txt	2014-09-12 00:46:32.000000000 +0200
+++ b/channels/CMakeLists.txt	2015-09-22 10:49:35.143095353 +0200
@@ -202,8 +202,8 @@
 file(GLOB FILEPATHS RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} "*/${FILENAME}")
 
 foreach(FILEPATH ${FILEPATHS})
-	if(${FILEPATH} MATCHES "^([^/]*)//${FILENAME}")
-		string(REGEX REPLACE "^([^/]*)//${FILENAME}" "\\1" DIR ${FILEPATH})
+	if(${FILEPATH} MATCHES "^([^/]*)/+${FILENAME}")
+		string(REGEX REPLACE "^([^/]*)/+${FILENAME}" "\\1" DIR ${FILEPATH})
 		set(CHANNEL_OPTION)
 		include(${FILEPATH})
 		if(${CHANNEL_OPTION})
--- a/client/CMakeLists.txt	2014-09-12 00:46:32.000000000 +0200
+++ b/client/CMakeLists.txt	2015-09-22 10:49:35.143095353 +0200
@@ -59,8 +59,8 @@
 file(GLOB FILEPATHS RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} "*/${FILENAME}")
 
 foreach(FILEPATH ${FILEPATHS})
-	if(${FILEPATH} MATCHES "^([^/]*)//${FILENAME}")
-		string(REGEX REPLACE "^([^/]*)//${FILENAME}" "\\1" FREERDP_CLIENT ${FILEPATH})
+	if(${FILEPATH} MATCHES "^([^/]*)/+${FILENAME}")
+		string(REGEX REPLACE "^([^/]*)/+${FILENAME}" "\\1" FREERDP_CLIENT ${FILEPATH})
 		set(FREERDP_CLIENT_ENABLED 0)
 		include(${FILEPATH})
 		if(FREERDP_CLIENT_ENABLED)
--- a/client/X11/xf_event.c	2014-09-12 00:46:32.000000000 +0200
+++ b/client/X11/xf_event.c	2015-09-22 10:49:58.903515567 +0200
@@ -558,6 +558,7 @@
 	if (event->xfocus.mode == NotifyWhileGrabbed)
 		XUngrabKeyboard(xfc->display, CurrentTime);
 
+	xf_keyboard_release_all_keypress(xfc);
 	xf_keyboard_clear(xfc);
 
 	if (app)
--- a/client/X11/xf_keyboard.c	2014-09-12 00:46:32.000000000 +0200
+++ b/client/X11/xf_keyboard.c	2015-09-22 10:49:58.903515567 +0200
@@ -169,6 +169,12 @@
 		if (xfc->KeyboardState[keycode] != NoSymbol)
 		{
 			rdp_scancode = freerdp_keyboard_get_rdp_scancode_from_x11_keycode(keycode);
+
+			// release tab before releasing the windows key.
+			// this stops the start menu from opening on unfocus event.
+			if (rdp_scancode == RDP_SCANCODE_LWIN)
+				freerdp_input_send_keyboard_event_ex(xfc->instance->input, FALSE, RDP_SCANCODE_TAB);
+
 			freerdp_input_send_keyboard_event_ex(xfc->instance->input, FALSE, rdp_scancode);
 			xfc->KeyboardState[keycode] = NoSymbol;
 		}
--- a/server/CMakeLists.txt	2014-09-12 00:46:32.000000000 +0200
+++ b/server/CMakeLists.txt	2015-09-22 10:49:35.143095353 +0200
@@ -44,8 +44,8 @@
 file(GLOB FILEPATHS RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} "*/${FILENAME}")
 
 foreach(FILEPATH ${FILEPATHS})
-	if(${FILEPATH} MATCHES "^([^/]*)//${FILENAME}")
-		string(REGEX REPLACE "^([^/]*)//${FILENAME}" "\\1" FREERDP_SERVER ${FILEPATH})
+	if(${FILEPATH} MATCHES "^([^/]*)/+${FILENAME}")
+		string(REGEX REPLACE "^([^/]*)/+${FILENAME}" "\\1" FREERDP_SERVER ${FILEPATH})
 		set(FREERDP_SERVER_ENABLED 0)
 		include(${FILEPATH})
 		if(FREERDP_SERVER_ENABLED)
--- a/third-party/CMakeLists.txt	2014-09-12 00:46:32.000000000 +0200
+++ b/third-party/CMakeLists.txt	2015-09-22 10:49:35.143095353 +0200
@@ -22,11 +22,11 @@
 file(GLOB all_valid_subdirs RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} "*/CMakeLists.txt")
 
 foreach(dir ${all_valid_subdirs})
-	if(${dir} MATCHES "^([^/]*)//CMakeLists.txt")
-		string(REGEX REPLACE "^([^/]*)//CMakeLists.txt" "\\1" dir_trimmed ${dir})
+	if(${dir} MATCHES "^([^/]*)/+CMakeLists.txt")
+		string(REGEX REPLACE "^([^/]*)/+CMakeLists.txt" "\\1" dir_trimmed ${dir})
 		message(STATUS "Adding third-party component ${dir_trimmed}")
 		add_subdirectory(${dir_trimmed})
 	endif()
 endforeach(dir)
 
-set(THIRD_PARTY_INCLUDES ${THIRD_PARTY_INCLUDES} PARENT_SCOPE)
\ No newline at end of file
+set(THIRD_PARTY_INCLUDES ${THIRD_PARTY_INCLUDES} PARENT_SCOPE)
