--- a/cmd/gvedit/csettings.cpp	2014-04-13 22:40:25.000000000 +0200
+++ b/cmd/gvedit/csettings.cpp	2015-09-20 13:04:00.806330578 +0200
@@ -16,7 +16,7 @@
 #include "csettings.h"
 #include "qmessagebox.h"
 #include "qfiledialog.h"
-#include <QtGui>
+#include <QtWidgets>
 #include <qfile.h>
 #include "mdichild.h"
 #include "string.h"
@@ -374,7 +374,7 @@
 		QString msg ("Output written to ");
 		msg.append(fileName);
 		msg.append("\n");
-		errorPipe((char *) msg.toAscii().constData());
+		errorPipe((char *) msg.toLatin1().constData());
 	    }
 	}
 
--- a/cmd/gvedit/imageviewer.h	2014-04-13 22:40:25.000000000 +0200
+++ b/cmd/gvedit/imageviewer.h	2015-09-20 13:03:58.782273993 +0200
@@ -15,7 +15,7 @@
 
 #ifndef IMAGEVIEWER_H
 #define IMAGEVIEWER_H
-#include <QtGui>
+#include <QtWidgets>
 #include <QMainWindow>
 #include <QPrinter>
 
--- a/cmd/gvedit/mainwindow.cpp	2014-04-13 22:40:25.000000000 +0200
+++ b/cmd/gvedit/mainwindow.cpp	2015-09-20 13:04:00.806330578 +0200
@@ -11,7 +11,7 @@
  * Contributors: See CVS logs. Details at http://www.graphviz.org/
  *************************************************************************/
 
-#include <QtGui>
+#include <QtWidgets>
 #include <qframe.h>
 #include "mainwindow.h"
 #include "mdichild.h"
@@ -289,7 +289,7 @@
 	msg.append("working on ");
 	msg.append(activeMdiChild()->currentFile());
 	msg.append("\n");
-	errorPipe((char *) msg.toAscii().constData());
+	errorPipe((char *) msg.toLatin1().constData());
 	prevChild = activeMdiChild();
     }
 }
--- a/cmd/gvedit/mdichild.cpp	2014-04-13 22:40:25.000000000 +0200
+++ b/cmd/gvedit/mdichild.cpp	2015-09-20 13:04:00.806330578 +0200
@@ -12,7 +12,7 @@
  *************************************************************************/
 
 
-#include <QtGui>
+#include <QtWidgets>
 
 #include "mdichild.h"
 #include "mainwindow.h"
--- a/cmd/gvedit/ui_settings.h	2014-04-13 22:40:25.000000000 +0200
+++ b/cmd/gvedit/ui_settings.h	2015-09-20 13:03:58.782273993 +0200
@@ -10,22 +10,22 @@
 #ifndef UI_SETTINGS_H
 #define UI_SETTINGS_H
 
-#include <QtCore/QVariant>
-#include <QtGui/QAction>
-#include <QtGui/QApplication>
-#include <QtGui/QButtonGroup>
-#include <QtGui/QComboBox>
-#include <QtGui/QDialog>
-#include <QtGui/QFrame>
-#include <QtGui/QHBoxLayout>
-#include <QtGui/QHeaderView>
-#include <QtGui/QLabel>
-#include <QtGui/QLineEdit>
-#include <QtGui/QPushButton>
-#include <QtGui/QSpacerItem>
-#include <QtGui/QTextEdit>
-#include <QtGui/QVBoxLayout>
-#include <QtGui/QWidget>
+#include <QVariant>
+#include <QAction>
+#include <QApplication>
+#include <QButtonGroup>
+#include <QComboBox>
+#include <QDialog>
+#include <QFrame>
+#include <QHBoxLayout>
+#include <QHeaderView>
+#include <QLabel>
+#include <QLineEdit>
+#include <QPushButton>
+#include <QSpacerItem>
+#include <QTextEdit>
+#include <QVBoxLayout>
+#include <QWidget>
 
 QT_BEGIN_NAMESPACE
 
@@ -343,49 +343,49 @@
 
     void retranslateUi(QDialog *Dialog)
     {
-        Dialog->setWindowTitle(QApplication::translate("Dialog", "Dialog", 0, QApplication::UnicodeUTF8));
-        actionCSettingsOK->setText(QApplication::translate("Dialog", "CSettingsOK", 0, QApplication::UnicodeUTF8));
-        label->setText(QApplication::translate("Dialog", "Layout Engine", 0, QApplication::UnicodeUTF8));
+        Dialog->setWindowTitle(QApplication::translate("Dialog", "Dialog", 0));
+        actionCSettingsOK->setText(QApplication::translate("Dialog", "CSettingsOK", 0));
+        label->setText(QApplication::translate("Dialog", "Layout Engine", 0));
         cbLayout->clear();
         cbLayout->insertItems(0, QStringList()
-         << QApplication::translate("Dialog", "dot", 0, QApplication::UnicodeUTF8)
-         << QApplication::translate("Dialog", "circo", 0, QApplication::UnicodeUTF8)
-         << QApplication::translate("Dialog", "fdp", 0, QApplication::UnicodeUTF8)
-         << QApplication::translate("Dialog", "neato", 0, QApplication::UnicodeUTF8)
-         << QApplication::translate("Dialog", "nop", 0, QApplication::UnicodeUTF8)
-         << QApplication::translate("Dialog", "nop1", 0, QApplication::UnicodeUTF8)
-         << QApplication::translate("Dialog", "nop2", 0, QApplication::UnicodeUTF8)
-         << QApplication::translate("Dialog", "osage", 0, QApplication::UnicodeUTF8)
-         << QApplication::translate("Dialog", "patchwork", 0, QApplication::UnicodeUTF8)
-         << QApplication::translate("Dialog", "sfdp", 0, QApplication::UnicodeUTF8)
-         << QApplication::translate("Dialog", "twopi", 0, QApplication::UnicodeUTF8)
+         << QApplication::translate("Dialog", "dot", 0)
+         << QApplication::translate("Dialog", "circo", 0)
+         << QApplication::translate("Dialog", "fdp", 0)
+         << QApplication::translate("Dialog", "neato", 0)
+         << QApplication::translate("Dialog", "nop", 0)
+         << QApplication::translate("Dialog", "nop1", 0)
+         << QApplication::translate("Dialog", "nop2", 0)
+         << QApplication::translate("Dialog", "osage", 0)
+         << QApplication::translate("Dialog", "patchwork", 0)
+         << QApplication::translate("Dialog", "sfdp", 0)
+         << QApplication::translate("Dialog", "twopi", 0)
         );
-        label_3->setText(QApplication::translate("Dialog", "Output File Type", 0, QApplication::UnicodeUTF8));
+        label_3->setText(QApplication::translate("Dialog", "Output File Type", 0));
         cbExtension->clear();
         cbExtension->insertItems(0, QStringList()
-         << QApplication::translate("Dialog", "png", 0, QApplication::UnicodeUTF8)
-         << QApplication::translate("Dialog", "jpg", 0, QApplication::UnicodeUTF8)
-         << QApplication::translate("Dialog", "ps", 0, QApplication::UnicodeUTF8)
-         << QApplication::translate("Dialog", "dot", 0, QApplication::UnicodeUTF8)
+         << QApplication::translate("Dialog", "png", 0)
+         << QApplication::translate("Dialog", "jpg", 0)
+         << QApplication::translate("Dialog", "ps", 0)
+         << QApplication::translate("Dialog", "dot", 0)
         );
-        label_2->setText(QApplication::translate("Dialog", "Output File Name", 0, QApplication::UnicodeUTF8));
-        pbOut->setText(QApplication::translate("Dialog", "...", 0, QApplication::UnicodeUTF8));
-        label_5->setText(QApplication::translate("Dialog", "Scope", 0, QApplication::UnicodeUTF8));
+        label_2->setText(QApplication::translate("Dialog", "Output File Name", 0));
+        pbOut->setText(QApplication::translate("Dialog", "...", 0));
+        label_5->setText(QApplication::translate("Dialog", "Scope", 0));
         cbScope->clear();
         cbScope->insertItems(0, QStringList()
-         << QApplication::translate("Dialog", "graph", 0, QApplication::UnicodeUTF8)
-         << QApplication::translate("Dialog", "node", 0, QApplication::UnicodeUTF8)
-         << QApplication::translate("Dialog", "edge", 0, QApplication::UnicodeUTF8)
+         << QApplication::translate("Dialog", "graph", 0)
+         << QApplication::translate("Dialog", "node", 0)
+         << QApplication::translate("Dialog", "edge", 0)
         );
-        label_6->setText(QApplication::translate("Dialog", "Name", 0, QApplication::UnicodeUTF8));
-        label_7->setText(QApplication::translate("Dialog", "Value", 0, QApplication::UnicodeUTF8));
-        pbAdd->setText(QApplication::translate("Dialog", "Add", 0, QApplication::UnicodeUTF8));
-        pbHelp->setText(QApplication::translate("Dialog", "Help", 0, QApplication::UnicodeUTF8));
-        pbNew->setText(QApplication::translate("Dialog", "clear", 0, QApplication::UnicodeUTF8));
-        pbOpen->setText(QApplication::translate("Dialog", "load", 0, QApplication::UnicodeUTF8));
-        pbSave->setText(QApplication::translate("Dialog", "save", 0, QApplication::UnicodeUTF8));
-        btnCancel->setText(QApplication::translate("Dialog", "Cancel", 0, QApplication::UnicodeUTF8));
-        btnOK->setText(QApplication::translate("Dialog", "OK", 0, QApplication::UnicodeUTF8));
+        label_6->setText(QApplication::translate("Dialog", "Name", 0));
+        label_7->setText(QApplication::translate("Dialog", "Value", 0));
+        pbAdd->setText(QApplication::translate("Dialog", "Add", 0));
+        pbHelp->setText(QApplication::translate("Dialog", "Help", 0));
+        pbNew->setText(QApplication::translate("Dialog", "clear", 0));
+        pbOpen->setText(QApplication::translate("Dialog", "load", 0));
+        pbSave->setText(QApplication::translate("Dialog", "save", 0));
+        btnCancel->setText(QApplication::translate("Dialog", "Cancel", 0));
+        btnOK->setText(QApplication::translate("Dialog", "OK", 0));
     } // retranslateUi
 
 };
--- a/configure.ac	2014-04-13 22:40:25.000000000 +0200
+++ b/configure.ac	2015-09-20 13:03:52.822106965 +0200
@@ -2613,10 +2613,10 @@
   use_qt="No (disabled)"
 else
 
-  AC_CHECK_PROGS(QMAKE,qmake-qt4 qmake-qt3 qmake,false)
+  AC_CHECK_PROGS(QMAKE,qmake-qt5 qmake,false)
   if test "$QMAKE" != "false"; then
-    PKG_CHECK_MODULES(QTCORE, [QtCore],[
-      PKG_CHECK_MODULES(QTGUI, [QtGui],[
+    PKG_CHECK_MODULES(QTCORE, [Qt5Core],[
+      PKG_CHECK_MODULES(QTGUI, [Qt5Widgets Qt5PrintSupport],[
 	use_qt="Yes"
       ],[
         use_qt="No (QtGui not available)"
