--- a/src/mesa/drivers/dri/i965/intel_screen.c	2015-04-09 13:26:33.000000000 +0200
+++ b/src/mesa/drivers/dri/i965/intel_screen.c	2015-04-16 20:31:51.361459748 +0200
@@ -1126,7 +1126,8 @@
 {
    static const mesa_format formats[] = {
       MESA_FORMAT_B5G6R5_UNORM,
-      MESA_FORMAT_B8G8R8A8_UNORM
+      MESA_FORMAT_B8G8R8A8_UNORM,
+      MESA_FORMAT_B8G8R8X8_UNORM
    };
 
    /* GLX_SWAP_COPY_OML is not supported due to page flipping. */
