--- a/data/kde.categories	2015-10-09 01:20:12.000000000 +0200
+++ b/data/kde.categories	2015-11-08 12:48:50.891106407 +0100
@@ -34,6 +34,14 @@
 powerdevil powerdevil
 log_user_manager user manager
 log_kioremote kioslave (kioremove)
+kscreen.kded kscreen (kded)
+kscreen libkscreen
+kscreen.edid libkscreen (edid)
+kscreen.fake libkscreen (fake backend)
+kscreen.qscreen libkscreen (qscreen backend)
+kscreen.xcb_helper libkscreen (xcb helper)
+kscreen.xrandr libkscreen (xrandr backend)
+kscreen.xrandr11 libkscreen (xrandr1.1 backend)
 ksmserver ksmserver
 libksysguard libksysguard
 log_kio_man kioslave (kio_man)
