--- a/camlibs/ax203/jpeg_memsrcdest.c	2015-08-08 15:57:29.000000000 +0000
+++ b/camlibs/ax203/jpeg_memsrcdest.c	2016-06-21 05:24:07.950374334 +0000
@@ -25,8 +25,8 @@
 #include "jpeg_memsrcdest.h"
 
 /* libjpeg8 and later come with their own (API compatible) memory source
-   and dest */
-#if JPEG_LIB_VERSION < 80
+   and dest, and older versions may have it backported */
+#if JPEG_LIB_VERSION < 80 && !defined(MEM_SRCDST_SUPPORTED)
 
 /* Expanded data source object for memory input */
 
--- a/camlibs/ax203/jpeg_memsrcdest.h	2015-08-08 15:57:29.000000000 +0000
+++ b/camlibs/ax203/jpeg_memsrcdest.h	2016-06-21 05:24:07.950374334 +0000
@@ -1,5 +1,7 @@
 #include <jpeglib.h>
 
+#if JPEG_LIB_VERSION < 80 && !defined(MEM_SRCDST_SUPPORTED)
+
 void
 jpeg_mem_src (j_decompress_ptr cinfo, unsigned char * buffer,
 	unsigned long bufsize);
@@ -7,3 +9,5 @@
 void
 jpeg_mem_dest (j_compress_ptr cinfo, unsigned char ** outbuffer,
 	unsigned long * outsize);
+
+#endif
--- a/camlibs/jl2005c/jpeg_memsrcdest.c	2015-08-08 15:57:29.000000000 +0000
+++ b/camlibs/jl2005c/jpeg_memsrcdest.c	2016-06-21 05:24:07.951374348 +0000
@@ -25,8 +25,8 @@
 #include "jpeg_memsrcdest.h"
 
 /* libjpeg8 and later come with their own (API compatible) memory source
-   and dest */
-#if JPEG_LIB_VERSION < 80
+   and dest, and older versions may have it backported */
+#if JPEG_LIB_VERSION < 80 && !defined(MEM_SRCDST_SUPPORTED)
 
 /* Expanded data source object for memory input */
 
--- a/camlibs/jl2005c/jpeg_memsrcdest.h	2015-08-08 15:57:29.000000000 +0000
+++ b/camlibs/jl2005c/jpeg_memsrcdest.h	2016-06-21 05:24:07.951374348 +0000
@@ -1,5 +1,7 @@
 #include <jpeglib.h>
 
+#if JPEG_LIB_VERSION < 80 && !defined(MEM_SRCDST_SUPPORTED)
+
 void
 jpeg_mem_src (j_decompress_ptr cinfo, unsigned char * buffer,
 	unsigned long bufsize);
@@ -7,3 +9,5 @@
 void
 jpeg_mem_dest (j_compress_ptr cinfo, unsigned char ** outbuffer,
 	unsigned long * outsize);
+
+#endif
