--- a/ytnef.c	2004-08-26 19:09:05.000000000 +0200
+++ b/ytnef.c	2016-07-25 23:24:40.447726452 +0200
@@ -565,7 +565,7 @@
 int TNEFHexBreakdown STD_ARGLIST {
     int i;
     if (TNEF->Debug == 0) 
-        return;
+        return -1;
 
     printf("%s: [%i bytes] \n", TNEFList[id].name, size);
 
@@ -580,7 +580,7 @@
 int TNEFDetailedPrint STD_ARGLIST {
     int i;
     if (TNEF->Debug == 0) 
-        return;
+        return -1;
 
     printf("%s: [%i bytes] \n", TNEFList[id].name, size);
 
@@ -1328,7 +1328,7 @@
 
     comp_Prebuf.size = strlen(RTF_PREBUF);
     comp_Prebuf.data = calloc(comp_Prebuf.size, 1);
-    strcpy(comp_Prebuf.data, RTF_PREBUF);
+    memcpy(comp_Prebuf.data, RTF_PREBUF, comp_Prebuf.size);
 
     src = p->data;
     in = 0;
