--- a/tools/llvm-shlib/CMakeLists.txt	2015-10-03 15:19:53.761598376 +0200
+++ b/tools/llvm-shlib/CMakeLists.txt	2015-10-03 16:00:56.673294348 +0200
@@ -64,7 +64,7 @@
 
   if (LLVM_DYLIB_EXPORT_ALL)
     add_custom_command(OUTPUT ${LLVM_EXPORTED_SYMBOL_FILE}
-      COMMAND echo \"LLVM*\" > ${LLVM_EXPORTED_SYMBOL_FILE} && echo \"_Z*llvm*\" >> ${LLVM_EXPORTED_SYMBOL_FILE}
+      COMMAND echo -e \"LLVM*\\n_Z*llvm*\\nConvertUTF*\\ngetNumBytesForUTF8\\nisLegalUTF8*\" > ${LLVM_EXPORTED_SYMBOL_FILE}
       WORKING_DIRECTORY ${LIB_DIR}
       DEPENDS ${LLVM_DYLIB_REQUIRED_EXPORTS}
       COMMENT "Generating combined export list...")
