diff -Naur a/build/openldap.m4 b/build/openldap.m4
--- a/build/openldap.m4	2014-10-02 08:58:54.952359661 +0200
+++ b/build/openldap.m4	2014-10-02 08:59:54.792454189 +0200
@@ -547,19 +547,13 @@
 #ifndef DB_VERSION_MINOR
 #	define DB_VERSION_MINOR 0
 #endif
-#ifndef DB_VERSION_PATCH
-#      define DB_VERSION_PATCH 0
-#endif
 
-#define DB_VERSION_FULL        ((DB_VERSION_MAJOR<<16)|(DB_VERSION_MINOR<<8)|DB_VERSION_PATCH)
+#define DB_VERSION_MM  ((DB_VERSION_MAJOR<<8)|DB_VERSION_MINOR)
 
-/* require 4.4 or later, but less than 6.0.20 */
-#if DB_VERSION_FULL >= 0x040400 && DB_VERSION_FULL < 0x060014
+/* require 4.4 or later */
+#if DB_VERSION_MM >= 0x0404
 	__db_version_compat
 #endif
-#if DB_VERSION_FULL >= 0x060014
-#error "BerkeleyDB 6.0.20+ license is incompatible with LDAP"
-#endif
 	], [ol_cv_bdb_compat=yes], [ol_cv_bdb_compat=no])])
 ])
 
