--- a/ext/standard/info.c	2015-12-16 11:42:03.000000000 +0100
+++ b/ext/standard/info.c	2016-01-04 12:42:05.539339545 +0100
@@ -870,7 +870,7 @@
 #ifdef ARCHITECTURE
 		php_info_print_table_row(2, "Architecture", ARCHITECTURE);
 #endif
-#ifdef CONFIGURE_COMMAND
+#if 0
 		php_info_print_table_row(2, "Configure Command", CONFIGURE_COMMAND );
 #endif
 
--- a/ext/standard/tests/general_functions/phpinfo.phpt	2015-12-16 11:42:01.000000000 +0100
+++ b/ext/standard/tests/general_functions/phpinfo.phpt	2016-01-04 12:42:05.539339545 +0100
@@ -20,7 +20,6 @@
 
 System => %s
 Build Date => %s%a
-Configure Command => %s
 Server API => Command Line Interface
 Virtual Directory Support => %s
 Configuration File (php.ini) Path => %s
