diff -Naur a/editor/connectioneditor.cpp b/editor/connectioneditor.cpp
--- a/editor/connectioneditor.cpp	2014-02-26 14:49:33.000000000 +0100
+++ b/editor/connectioneditor.cpp	2014-07-04 22:28:20.804856199 +0200
@@ -117,7 +117,7 @@
     action = new QAction(i18n("VLAN"), this);
     action->setData(NetworkManager::ConnectionSettings::Vlan);
     m_menu->addAction(action);
-#if NM_CHECK_VERSION(0, 9, 9)
+#if 0
     action = new QAction(i18n("Team"), this);
     action->setData(NetworkManager::ConnectionSettings::Team);
     m_menu->addAction(action);
diff -Naur a/lib/editor/CMakeLists.txt b/lib/editor/CMakeLists.txt
--- a/lib/editor/CMakeLists.txt	2014-02-26 14:49:33.000000000 +0100
+++ b/lib/editor/CMakeLists.txt	2014-07-04 22:28:20.804856199 +0200
@@ -29,12 +29,6 @@
     connectionwidget.cpp
 )
 
-if (${NETWORKMANAGERQT_VERSION} VERSION_GREATER 0.9.8.9)
-set(plasmanetworkmanagement_libeditor_SRCS
-    ${plasmanetworkmanagement_libeditor_SRCS}
-    teamwidget.cpp)
-endif()
-
 kde4_add_ui_files(plasmanetworkmanagement_libeditor_SRCS
     ui/advancedpermissionswidget.ui
     ui/ipv4.ui
@@ -60,10 +54,6 @@
     ui/connectionwidget.ui
 )
 
-if (${NETWORKMANAGERQT_VERSION} VERSION_GREATER 0.9.8.9)
-    kde4_add_ui_files(plasmanetworkmanagement_libeditor_SRCS ui/team.ui)
-endif()
-
 kde4_add_library(plasmanetworkmanagement-editor SHARED ${plasmanetworkmanagement_libeditor_SRCS})
 target_link_libraries(plasmanetworkmanagement-editor
     ${KDE4_KDEUI_LIBS}
diff -Naur a/lib/editor/connectiondetaileditor.cpp b/lib/editor/connectiondetaileditor.cpp
--- a/lib/editor/connectiondetaileditor.cpp	2014-02-26 14:49:33.000000000 +0100
+++ b/lib/editor/connectiondetaileditor.cpp	2014-07-04 22:29:05.347672960 +0200
@@ -38,7 +38,7 @@
 #include <bridgewidget.h>
 #include <vlanwidget.h>
 #include <wimaxwidget.h>
-#if NM_CHECK_VERSION(0, 9, 9)
+#if 0
 #include <teamwidget.h>
 #endif
 #include "vpnuiplugin.h"
@@ -314,7 +314,7 @@
     } else if (type == NetworkManager::ConnectionSettings::Vlan) { // Vlan
         VlanWidget * vlanWidget = new VlanWidget(m_connection->setting(NetworkManager::Setting::Vlan), this);
         m_ui->tabWidget->addTab(vlanWidget, i18n("Vlan"));
-#if NM_CHECK_VERSION(0, 9, 9)
+#if 0
     } else if (type == NetworkManager::ConnectionSettings::Team) { // Team
         TeamWidget * teamWidget = new TeamWidget(uuid, m_connection->setting(NetworkManager::Setting::Team), this);
         m_ui->tabWidget->addTab(teamWidget, i18n("Team"));
@@ -369,7 +369,7 @@
          || type == ConnectionSettings::Wireless
          || type == ConnectionSettings::Infiniband
          || type == ConnectionSettings::Wimax
-     #if NM_CHECK_VERSION(0, 9, 9)
+     #if 0
          || type == ConnectionSettings::Team
      #endif
          || type == ConnectionSettings::Bond
