--- a/setup.py	2015-12-05 20:47:20.000000000 +0100
+++ b/setup.py	2016-04-16 21:57:49.749811759 +0200
@@ -885,7 +885,7 @@
         # a release.  Most open source OSes come with one or more
         # versions of BerkeleyDB already installed.
 
-        max_db_ver = (5, 3)
+        max_db_ver = (6, 2)
         min_db_ver = (4, 3)
         db_setup_debug = False   # verbose debug prints from this script?
 
@@ -907,7 +907,11 @@
             return True
 
         def gen_db_minor_ver_nums(major):
-            if major == 5:
+            if major == 6:
+                for x in range(max_db_ver[1]+1):
+                    if allow_db_ver((6, x)):
+                        yield x
+            elif major == 5:
                 for x in range(max_db_ver[1]+1):
                     if allow_db_ver((5, x)):
                         yield x
