--- a/setup.py	2015-12-07 02:39:11.000000000 +0100
+++ b/setup.py	2016-04-16 21:59:00.054050469 +0200
@@ -874,7 +874,7 @@
         # a release.  Most open source OSes come with one or more
         # versions of BerkeleyDB already installed.
 
-        max_db_ver = (5, 3)
+        max_db_ver = (6, 2)
         min_db_ver = (3, 3)
         db_setup_debug = False   # verbose debug prints from this script?
 
@@ -889,7 +889,15 @@
             return True
 
         def gen_db_minor_ver_nums(major):
-            if major == 4:
+            if major == 6:
+                for x in range(max_db_ver[1]+1):
+                    if allow_db_ver((6, x)):
+                        yield x
+            elif major == 5:
+                for x in range(max_db_ver[1]+1):
+                    if allow_db_ver((5, x)):
+                        yield x
+            elif major == 4:
                 for x in range(max_db_ver[1]+1):
                     if allow_db_ver((4, x)):
                         yield x
