--- a/plugins/qca-ossl/qca-ossl.cpp	2015-10-02 11:39:21.000000000 +0200
+++ b/plugins/qca-ossl/qca-ossl.cpp	2016-03-03 20:54:52.087065857 +0100
@@ -5403,9 +5403,11 @@
 			ctx = SSL_CTX_new(SSLv2_client_method());
 			break;
 #endif
+#ifndef OPENSSL_NO_SSL3_METHOD
 		case TLS::SSL_v3:
 			ctx = SSL_CTX_new(SSLv3_client_method());
 			break;
+#endif
 		case TLS::TLS_v1:
 			ctx = SSL_CTX_new(TLSv1_client_method());
 			break;
