--- a/Python/configure.py	2015-04-20 16:38:24.000000000 +0200
+++ b/Python/configure.py	2015-05-29 10:38:32.876611996 +0200
@@ -32,6 +32,8 @@
 import optparse
 import sys
 
+src_dir = os.path.dirname(os.path.abspath(__file__))
+
 
 ###############################################################################
 # You shouldn't need to modify anything above this line.
@@ -206,6 +208,8 @@
                     "The QScintilla version number could not be determined by "
                     "reading %s." % sciglobal)
 
+        return # Debian: do not check for the installed version, we're good this way.
+
         lib_dir = target_configuration.qsci_lib_dir
         if lib_dir is None:
             lib_dir = target_configuration.qt_lib_dir
@@ -260,7 +264,12 @@
         the target configuration.
         """
 
-        return 'sip/qscimod5.sip' if target_configuration.pyqt_package == 'PyQt5' else 'sip/qscimod4.sip'
+        if target_configuration.pyqt_package == 'PyQt5':
+            return os.path.join(src_dir, 'sip/qscimod5.sip')
+        else:
+            return os.path.join(src_dir, 'sip/qscimod4.sip')
+
+        #return 'sip/qscimod5.sip' if target_configuration.pyqt_package == 'PyQt5' else 'sip/qscimod4.sip'
 
     def get_sip_installs(self, target_configuration):
         """ Return a tuple of the installation directory of the module's .sip
