--- a/qtwebkit/Source/JavaScriptCore/runtime/JSObject.cpp	2015-02-17 05:57:11.000000000 +0100
+++ b/qtwebkit/Source/JavaScriptCore/runtime/JSObject.cpp	2015-04-10 09:22:01.141481012 +0200
@@ -1909,6 +1909,11 @@
     }
 }
 
+// Used in JSArray.cpp so we must instantiate explicit
+template void JSObject::putByIndexBeyondVectorLengthWithoutAttributes<Int32Shape>(ExecState* exec, unsigned i, JSValue value);
+template void JSObject::putByIndexBeyondVectorLengthWithoutAttributes<DoubleShape>(ExecState* exec, unsigned i, JSValue value);
+template void JSObject::putByIndexBeyondVectorLengthWithoutAttributes<ContiguousShape>(ExecState* exec, unsigned i, JSValue value);
+
 void JSObject::putByIndexBeyondVectorLengthWithArrayStorage(ExecState* exec, unsigned i, JSValue value, bool shouldThrow, ArrayStorage* storage)
 {
     VM& vm = exec->vm();
