--- a/src/core/config/desktop_linux.pri	2015-12-14 15:27:24.000000000 +0000
+++ b/src/core/config/desktop_linux.pri	2016-01-25 20:35:29.678185672 +0000
@@ -7,4 +7,5 @@
     enable_plugins=1 \
 
 linux-clang: GYP_CONFIG += clang=1 host_clang=1 clang_use_chrome_plugins=0 make_clang_dir=/usr
+linux-clang-libc++: GYP_CONFIG += clang=1 host_clang=1 clang_use_chrome_plugins=0 make_clang_dir=/usr use_libcpp=1
 else: GYP_CONFIG += clang=0 host_clang=0
--- a/tools/qmake/mkspecs/features/functions.prf	2015-12-14 15:27:24.000000000 +0000
+++ b/tools/qmake/mkspecs/features/functions.prf	2016-01-25 20:38:14.642821866 +0000
@@ -1,5 +1,5 @@
 defineTest(isPlatformSupported) {
-  !linux-g++*:!linux-clang:!win32-msvc2013*:!win32-msvc2015*:!macx-clang*:!boot2qt {
+  !linux-g++*:!linux-clang*:!win32-msvc2013*:!win32-msvc2015*:!macx-clang*:!boot2qt {
     skipBuild("Qt WebEngine can currently only be built for Linux (GCC/clang), Windows (MSVC 2013 or 2015), OS X (10.9/XCode 5.1+) or Qt for Device Creation.")
     return(false)
   }
