--- a/sound/pci/hda/patch_conexant.c	2016-03-10 11:37:29.918081515 +0100
+++ b/sound/pci/hda/patch_conexant.c	2016-03-10 11:37:59.257047226 +0100
@@ -901,7 +901,8 @@
 		snd_hda_pick_fixup(codec, cxt5051_fixup_models,
 				   cxt5051_fixups, cxt_fixups);
 		break;
-	case 0x14f150f2:
+	case 0x14f150f2: /* CX20722 */
+	case 0x14f150f4: /* CX20724 */
 		codec->power_save_node = 1;
 		/* Fall through */
 	default:
