--- a/unix/xserver/hw/vnc/Input.c	2015-12-24 21:18:02.000000000 +0100
+++ b/unix/xserver/hw/vnc/Input.c	2016-05-26 12:35:04.845341097 +0200
@@ -300,6 +300,8 @@
 #if XORG < 111
 	n = GetKeyboardEvents(eventq, dev, action, kc);
 	enqueueEvents(dev, n);
+#elif XORG > 117
+	QueueKeyboardEvents(dev, action, kc);
 #else
 	QueueKeyboardEvents(dev, action, kc, NULL);
 #endif
--- a/unix/xserver/hw/vnc/Makefile.am	2015-12-24 21:18:02.000000000 +0100
+++ b/unix/xserver/hw/vnc/Makefile.am	2016-05-26 12:35:05.081343562 +0200
@@ -5,6 +5,7 @@
 RDR_LIB=$(LIB_DIR)/rdr/librdr.la
 NETWORK_LIB=$(LIB_DIR)/network/libnetwork.la
 XREGION_LIB=$(LIB_DIR)/Xregion/libXregion.la
+OS_LIB=$(LIB_DIR)/os/libos.la
 COMMON_LIBS=$(NETWORK_LIB) $(RFB_LIB) $(RDR_LIB) $(XREGION_LIB)
 
 noinst_LTLIBRARIES = libvnccommon.la
@@ -64,6 +65,6 @@
 
 libvnc_la_LDFLAGS = -module -avoid-version -Wl,-z,now
 
-libvnc_la_LIBADD = libvnccommon.la $(COMMON_LIBS)
+libvnc_la_LIBADD = libvnccommon.la $(COMMON_LIBS) $(OS_LIB)
 
 EXTRA_DIST = Xvnc.man
--- a/unix/xserver/hw/vnc/xorg-version.h	2015-12-24 21:18:02.000000000 +0100
+++ b/unix/xserver/hw/vnc/xorg-version.h	2016-05-26 12:35:04.845341097 +0200
@@ -48,8 +48,10 @@
 #define XORG 116
 #elif XORG_VERSION_CURRENT < ((1 * 10000000) + (17 * 100000) + (99 * 1000))
 #define XORG 117
+#elif XORG_VERSION_CURRENT < ((1 * 10000000) + (18 * 100000) + (99 * 1000))
+#define XORG 118
 #else
-#error "X.Org newer than 1.17 is not supported"
+#error "X.Org newer than 1.18 is not supported"
 #endif
 
 #endif
