--- a/vinagre/vinagre-bookmarks-ui.c	2013-07-02 16:14:35.000000000 +0200
+++ b/vinagre/vinagre-bookmarks-ui.c	2016-01-21 22:42:31.904955872 +0100
@@ -327,6 +327,7 @@
 				   title);
 
   gtk_message_dialog_format_secondary_markup (GTK_MESSAGE_DIALOG (dialog),
+					      "%s %s",
 					      msg2,
 					      name);
  
--- a/vinagre/vinagre-utils.vala	2013-07-23 16:24:34.000000000 +0200
+++ b/vinagre/vinagre-utils.vala	2016-01-21 22:40:20.928383352 +0100
@@ -109,7 +109,9 @@
 
         var auth_label = xml.get_object ("auth_required_label") as Label;
         // Translators: %s is a protocol, like VNC or SSH
-        auth_label.label = _("%s authentication is required").printf (protocol);
+        var auth_label_text = _("authentication is required");
+
+        auth_label.label = "%s %s".printf (protocol, auth_label_text);
 
         var host_label = xml.get_object ("host_label") as Label;
         host_label.label = host;
