diff -Naur a/libzeitgeist/log.vala b/libzeitgeist/log.vala
--- a/libzeitgeist/log.vala	2014-01-27 22:45:42.000000000 +0100
+++ b/libzeitgeist/log.vala	2014-09-18 22:25:35.449445776 +0200
@@ -40,6 +40,7 @@
  * Zeitgeist also comes with a blacklist extension to make sure the user
  * always stays in control of what information is logged.
  */
+
 namespace Zeitgeist
 {
 
@@ -58,8 +59,8 @@
  */
 public class Log : QueuedProxyWrapper
 {
-    [CCode (cheader_filename = "sys/sysinfo.h", cname = "get_nprocs_conf")]
-    extern static int get_nprocs_conf ();
+    [CCode (cname = "g_get_num_processors")]
+    private extern static uint get_num_processors ();
 
     class DbWorker
     {
@@ -141,7 +142,7 @@
         try {
             threads = new ThreadPool<DbWorker>.with_owned_data ((worker) => {
                 worker.run ();
-            }, get_nprocs_conf (), true);
+            }, (int) get_num_processors (), true);
         } catch (ThreadError err) {
             warning ("%s", err.message);
             threads = null;
@@ -637,4 +638,3 @@
 
 }
 
-// vim:expandtab:ts=4:sw=4
diff -Naur a/libzeitgeist/where-clause.vala b/libzeitgeist/where-clause.vala
--- a/libzeitgeist/where-clause.vala	2014-01-27 22:45:42.000000000 +0100
+++ b/libzeitgeist/where-clause.vala	2014-09-18 22:25:46.202837269 +0200
@@ -211,7 +211,11 @@
          */
         private T[] generic_array_to_unowned_array<T> (GenericArray<T> gptrarr)
         {
+#if VALA_0_24
+            long[] pointers = new long[gptrarr.length];
+#else
             long[] pointers = new long[gptrarr.length + 1];
+#endif
             Memory.copy(pointers, ((PtrArray *) gptrarr)->pdata,
                 (gptrarr.length) * sizeof (void *));
             return (T[]) pointers;
diff -Naur a/libzeitgeist/zeitgeist-2.0.deps b/libzeitgeist/zeitgeist-2.0.deps
--- a/libzeitgeist/zeitgeist-2.0.deps	2014-01-27 22:45:42.000000000 +0100
+++ b/libzeitgeist/zeitgeist-2.0.deps	2014-09-18 22:25:22.526040981 +0200
@@ -1 +1,2 @@
 zeitgeist-datamodel-2.0
+gio-2.0
