--- a/src/backend.cpp	2016-04-18 04:25:24.000000000 +0000
+++ b/src/backend.cpp	2016-07-25 23:42:22.901440435 +0000
@@ -96,7 +96,7 @@
                         qApp->applicationVersion(),
                         PHONON_VERSION_STR,
                         PHONON_VLC_VERSION);
-            libvlc_set_user_agent(libvlc,
+            libvlc_set_user_agent(pvlc_libvlc,
                                   qApp->applicationName().toUtf8().constData(),
                                   userAgent.toUtf8().constData());
         } else {
@@ -121,7 +121,7 @@
                 // If we failed to get a proper icon name, use the appname instead.
                 icon = qApp->applicationName().toLower();
             }
-            libvlc_set_app_id(libvlc,
+            libvlc_set_app_id(pvlc_libvlc,
                               id.toUtf8().constData(),
                               version.toUtf8().constData(),
                               icon.toUtf8().constData());
@@ -176,7 +176,7 @@
 
 QObject *Backend::createObject(BackendInterface::Class c, QObject *parent, const QList<QVariant> &args)
 {
-    if (!LibVLC::self || !libvlc)
+    if (!LibVLC::self || !pvlc_libvlc)
         return 0;
 
     switch (c) {
--- a/src/devicemanager.cpp	2016-04-18 04:25:24.000000000 +0000
+++ b/src/devicemanager.cpp	2016-07-25 23:42:22.901440435 +0000
@@ -203,7 +203,7 @@
 {
     QList<DeviceInfo> newDeviceList;
 
-    if (!LibVLC::self || !libvlc)
+    if (!LibVLC::self || !pvlc_libvlc)
         return;
 
     QList<QByteArray> audioOutBackends = vlcAudioOutBackends();
@@ -252,7 +252,7 @@
         bool hasDevices = false;
         VLC_FOREACH(audio_output_device,
                     device,
-                    libvlc_audio_output_device_list_get(libvlc, soundSystem),
+                    libvlc_audio_output_device_list_get(pvlc_libvlc, soundSystem),
                     libvlc_audio_output_device_list_release) {
             QString idName = QString::fromUtf8(device->psz_device);
             QString longName = QString::fromUtf8(device->psz_description);
@@ -275,11 +275,11 @@
             newDeviceList.append(info);
         }
 #else
-        const int deviceCount = libvlc_audio_output_device_count(libvlc, soundSystem);
+        const int deviceCount = libvlc_audio_output_device_count(pvlc_libvlc, soundSystem);
 
         for (int i = 0; i < deviceCount; i++) {
             VString idName(libvlc_audio_output_device_id(libvlc, soundSystem, i));
-            VString longName(libvlc_audio_output_device_longname(libvlc, soundSystem, i));
+            VString longName(libvlc_audio_output_device_longname(pvlc_libvlc, soundSystem, i));
 
             debug() << "found device" << soundSystem << idName << longName;
 
--- a/src/effectmanager.cpp	2016-04-18 04:25:24.000000000 +0000
+++ b/src/effectmanager.cpp	2016-07-25 23:42:22.902440425 +0000
@@ -46,7 +46,7 @@
 EffectManager::EffectManager(QObject *parent)
     : QObject(parent)
 {
-    if (!libvlc)
+    if (!pvlc_libvlc)
         return;
 
     updateEffects();
--- a/src/media.cpp	2016-04-18 04:25:24.000000000 +0000
+++ b/src/media.cpp	2016-07-25 23:42:22.902440425 +0000
@@ -30,7 +30,7 @@
 
 Media::Media(const QByteArray &mrl, QObject *parent) :
     QObject(parent),
-    m_media(libvlc_media_new_location(libvlc, mrl.constData())),
+    m_media(libvlc_media_new_location(pvlc_libvlc, mrl.constData())),
     m_mrl(mrl)
 {
     Q_ASSERT(m_media);
--- a/src/mediaplayer.cpp	2016-04-18 04:25:24.000000000 +0000
+++ b/src/mediaplayer.cpp	2016-07-25 23:42:22.902440425 +0000
@@ -50,7 +50,7 @@
 MediaPlayer::MediaPlayer(QObject *parent)
     : QObject(parent)
     , m_media(0)
-    , m_player(libvlc_media_player_new(libvlc))
+    , m_player(libvlc_media_player_new(pvlc_libvlc))
     , m_doingPausedPlay(false)
     , m_volume(75)
     , m_fadeAmount(1.0f)
--- a/src/utils/libvlc.h	2016-04-18 04:25:24.000000000 +0000
+++ b/src/utils/libvlc.h	2016-07-25 23:42:22.902440425 +0000
@@ -30,7 +30,7 @@
  * Please note that init() must have been called whenever using this, as no
  * checking of self is conducted (i.e. can be null).
  */
-#define libvlc LibVLC::self->vlc()
+#define pvlc_libvlc LibVLC::self->vlc()
 
 /**
  * Foreach loop macro for VLC descriptions.
@@ -50,7 +50,7 @@
 
 // This foreach expects only a type and variable because getter and releaser are generic.
 // Also the type is in short form i.e. libvlc_foo_t would be foo.
-#define VLC_FOREACH_LIST(type, variable) VLC_FOREACH(type, variable, libvlc_##type##_list_get(libvlc), libvlc_##type##_list_release)
+#define VLC_FOREACH_LIST(type, variable) VLC_FOREACH(type, variable, libvlc_##type##_list_get(pvlc_libvlc), libvlc_##type##_list_release)
 
 // These foreach expect no type because the type is generic, they do however
 // expect a getter to allow usage with our wrapper classes and since the getter
